### ISSUE TITLE
how can i add ids also to the pickle file ?

### ISSUE THREAD

**Issue by mathshangw** [open-mmlab/mmdetection#5267]
i'm new to use mmdetection to get the vectors for coco images i got result.pkl file with vectors for each image only not with its ids. How can i alter in the code to make the pickle file contains id of the image with its vector

**Comment by AronLin** [comment]
The test function is in this file - mmdet/apis/test.py.
You can do what you want here:

        result = model(return_loss=False, rescale=True, **data)
            # encode mask results
            if isinstance(result[0], tuple):
                result = [(bbox_results, encode_mask_results(mask_results))
                          for bbox_results, mask_results in result]
        ##Do what you want to do##
        ##Id can be obtained in data['img_metas'][0]['ori_filename']##
        results.extend(result)

**Comment by mathshangw** [comment]
thanks a lot for replying but the model i'm using used the class that i mentioned  in the post .. where exactly i can edit in the code ?

**Comment by AronLin** [comment]
The place you can edit is where I write comments.  
You can obtain the image id according to the file name.  
The origin result is a list, so you should extract its elements, build maps from image ids to the elements and repackage it to a list.

**Comment by mathshangw** [comment]
appreciate your replying but excuse me how can i obtain image id according to the file name. i did the extraction now and need to map ids to the result. 
Thanks a lot for your time

**Comment by mathshangw** [comment]
thanks a lot but extraction of elements will be by using pickle.dump .. like that 
`        pickle.dump([x for x, _ in results])` ?

**Comment by mathshangw** [comment]
thanks a lot i tried to get the id of the image by `img_meta['ori_filename']` and tried this code 
        ` data1=[]`
        `data1.append(img_meta['ori_filename'])`
        `f=open('file.txt','w')`
    `    for i in data1: ` 
             `f.write(i)`
but the file has one image only

**Comment by mathshangw** [comment]
excuse me i did the code as follow 

` if isinstance(result[0], tuple):`
           ` result = [(bbox_results, encode_mask_results(mask_results))`
                              `for bbox_results, mask_results in result]`
        `data1.append(img_meta['ori_filename'])  `
       ` images=[]`
       ` f=open('file.txt','w')`
      `  for i in data1:  `
          `  i=re.sub('.jpg', '', i) `
           ` i=i.replace("COCO_val2014_", '')  `   
           ` f.write(i+os.linesep)`
           ` images.append(i)`
       ` for _ in range(batch_size):`
            `prog_bar.update()`
       ` d={}`
       ` d= dict(zip(images, result))` 
       ` with open('filename.pkl','wb') as f1:`
                `pickle.dump(d, f1)`      
       
        `results.extend(result)`

and got the image_names and tried to make a dictionary for the result but the vectors i got is different from the original one i have already before adding image_names .. what is the wrong i did ?

**Comment by AronLin** [comment]
The function is for collecting results and they will be saved after multi-GPU collecting, so you should not save the results here.
Too many codes are not needed. Codes below is enough for val2017:
        
        with torch.no_grad():
            result = model(return_loss=False, rescale=True, **data)
            # encode mask results
            if isinstance(result[0], tuple):
                result = [(bbox_results, encode_mask_results(mask_results))
                          for bbox_results, mask_results in result]
        img_id = int(data['img_metas'][0].data[0][0]['ori_filename'].split('.')[0])
        result = [{img_id: result[0]}]
        results.extend(result)

For val2014, there is a little difference:

        img_id = int(data['img_metas'][0].data[0][0]['ori_filename'][13:].split('.')[0])

**Comment by mathshangw** [comment]
thanks a lot for your help .. i did it already now .. the vectors are right but i got the id for the first vector only like that 
`[{391895: [array([[3.43202576e+02, 2.43713760e+01, 4.96343414e+02, 3.32774170e+02,
        8.81948233e-01], dtype=float32), array([[3.44345245e+02, 4.28607712e+01, 4.91229034e+02, 3.36689850e+02,`

the problem now in each array i should have the id of it :-(

**Comment by mathshangw** [comment]
will the code be 

           `# encode mask results
             if isinstance(result[0], tuple):
             result = [(bbox_results, encode_mask_results(mask_results))
                      for bbox_results, mask_results in result]
    img_id = int(data['img_metas'][0].data[0][0]['ori_filename'].split('.')[0])
    result = [{img_id: result[0]}]
    results.extend(result)`

or 

       `# encode mask results
        if isinstance(result[0], tuple):
            result = [(bbox_results, encode_mask_results(mask_results))
                      for bbox_results, mask_results in result]
            img_id = int(data['img_metas'][0].data[0][0]['ori_filename'].split('.')[0])
            result = [{img_id: result[0]}]
    results.extend(result)`

**Comment by AronLin** [comment]
Please clarify the question, I don't know what you want to do. 
The result of an image is a list of N arrays and N means the number of classes. So the ids of N classes in an image should be the same.
So, I don't know why the ids you want in a result list of an image are different.
Or you can format the results and show me what you want.

**Comment by mathshangw** [comment]
the model i used should give for each image a vector or an array with its id so the result should be like that 
`{'266546173': array([[1.15911709e-03, 3.19864594e-05, 1.06177040e-05, 6.39094651e-07,
      dtype=float32), '150582765': array([[8.16741449e-05, 5.62177138e-06,`

what i got is the id for the first image only

**Comment by AronLin** [comment]
The origin output of an image is a list with N classes and each class has a bbox-score array. I showed you the method to map the image id to the list.
You mean you only need the bounding boxes and their scores of an image without their classes? 
Or you need each array of a class to have its image id?

**Comment by mathshangw** [comment]
You mean that the file I have now for one image with its classes so the id is the same right .. ?
But the size of the file is 290 MB .. I supposed that testing process by this command 

`python test.py /retinanet_fpn_coco_6401/latest.pth --show-dir /results.pkl --eval bbox`

Work on all val images in coco ( 40500 images) so the result will be for all of these images

**Comment by AronLin** [comment]
Did you run my code? 
The file `results.pkl` has a list and its length is the number of images.
Each element of the list is a map `{ image_id: class list}`. The class list has the length of N which is the number of classes. Each element of the class list is an array of boxes and scores for a class. Is it what you need? Or please clarify what you need.

The left problem is not about MMdet anymore. If you don't mind, please send emails to me directly. I'm glad to answer your question.
My email is 347630870@qq.com.

**Comment by mathshangw** [comment]
Thanks a lot ... can you check your mail please
